File: src/main/java/io/github/alexiscomete/lapinousecond/save/SaveLocation.java
package io.github.alexiscomete.lapinousecond.save;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Scanner;
import java.util.function.Function;

/**
 * @param <E> type of content, think to add a toString() method in E!
 */
public class SaveLocation<E> {

    public final static String pathStatic = getPathStatic();

    public static String getPathStatic() {
        return new File("").getAbsolutePath();
    }

    String sep;
    ArrayList<E> content = new ArrayList<>();
    String path;
    File file;
    Function<String, E> a;

    public SaveLocation(String separator, String path, Function<String, E> a) throws IOException {
        this.sep = separator;
        this.path = pathStatic + path;
        this.file = new File(pathStatic + path);
        this.a = a;
        if (path.endsWith("/")) {
            System.out.println(file.mkdirs());
        } else if (!file.exists()) {
            System.out.println(file.createNewFile());
        }
    }

    public void saveAll() {
        StringBuilder save = new StringBuilder();
        for (E e : content) {
            save.append(e.toString());
        }
        FileOutputStream fos;
        try {
            fos = new FileOutputStream(file);
            fos.write(String.valueOf(save).getBytes());
            fos.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void loadAll() {
        Scanner sc;
        try {
            sc = new Scanner(file);
            StringBuilder answer = new StringBuilder();
            while (sc.hasNextLine()) {
                answer.append(sc.nextLine());
                if (sc.hasNextLine()) answer.append("\n");
            }
            String[] str = String.valueOf(answer).split(sep);
            for (String s : str) {
                content.add(a.apply(s));
            }
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
    }

    public ArrayList<E> getContent() {
        return content;
    }

    public void setContent(ArrayList<E> content) {
        this.content = content;
    }

    public static void create(String path) throws IOException {
        File file = new File(pathStatic + path);
        if (path.endsWith("/")) {
            System.out.println(file.mkdirs());
        } else if (!file.exists()) {
            System.out.println(file.createNewFile());
        }
    }
}

File: src/main/java/io/github/alexiscomete/lapinousecond/useful/managesave/SaveLocation.kt
package io.github.alexiscomete.lapinousecond.useful.managesave

import java.io.File
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException
import java.util.*

@Synchronized
fun generateUniqueID(): Long {
    return try {
        Thread.sleep(2)
        System.currentTimeMillis()
    } catch (e: InterruptedException) {
        e.printStackTrace()
        System.currentTimeMillis()
    }
}

val pathStatic: String = File("").absolutePath

@Throws(IOException::class)
fun createPath(path: String) {
    val file = File(pathStatic + path)
    if (path.endsWith("/")) {
        println(file.mkdirs())
    } else if (!file.exists()) {
        println(file.createNewFile())
    }
}

/**
 * @param <E> type of content, think to add a toString() method in E!
</E> */
class SaveLocation<E>(private val sep: String, path: String, a: (String) -> E) {
    var content = ArrayList<E>()
    private val path: String
    private val file: File
    val a: (String) -> E

    init {
        this.path = pathStatic + path
        println("Searching file ${this.path}... ")
        file = File(pathStatic + path)
        this.a = a
        if (path.endsWith("/")) {
            println(file.mkdirs())
        } else if (!file.exists()) {
            println(file.createNewFile())
        }
    }

    fun saveAll() {
        val save = StringBuilder()
        for (e in content) {
            save.append(e.toString())
        }
        val fos: FileOutputStream
        try {
            fos = FileOutputStream(file)
            fos.write(save.toString().toByteArray())
            fos.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun loadAll() {
        val sc: Scanner
        try {
            sc = Scanner(file)
            val answer = StringBuilder()
            while (sc.hasNextLine()) {
                answer.append(sc.nextLine())
                if (sc.hasNextLine()) answer.append("\n")
            }
            val str = answer.toString().split(sep.toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            for (s in str) {
                content.add(a(s))
            }
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        }
    }
}